bundles/org.eclipse.mita.program.ui/src/org/eclipse/mita/program/ui/contentassist/ProgramDslProposalProvider.xtend

```diff
@@ -132,10 +144,110 @@ class ProgramDslProposalProvider extends AbstractProgramDslProposalProvider {
 	override getPriorityHelper() {
 		priorityHelper
 	}
+	
+	def proposeComplexTypeConstructors(Boolean scoped, EObject model, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
+		val scope = scopeProvider.getScope(model, ExpressionsPackage.eINSTANCE.elementReferenceExpression_Reference);
+		for (element : scope.allElements) {
+			val obj = element.EObjectOrProxy;
+			val fromPlatform = EcoreUtil2.getContainerOfType(obj, Program) === null;
+			val inSetupBlock = EcoreUtil2.getContainerOfType(model, SystemResourceSetup) !== null;
+			// in setup blocks we use platform defined types and vice versa
+			if(fromPlatform === inSetupBlock) {
+				if(obj instanceof SumAlternative || obj instanceof StructureType) {
+					val prefix = if(scoped && obj instanceof SumAlternative) {
+						(obj.eContainer as SumType).name + ".";
+					}
+					else {
+						"";
+					}
+					val s = completeTypeConstructor(prefix, obj as ComplexType);
+					if(s !== null) {
+						val proposal = createCompletionProposal(
+							s,
+							new StyledString(s),
+							labelProvider.getImage(obj),
+							context
+						);
+						
+						if (proposal instanceof ConfigurableCompletionProposal) {
+							proposal.additionalProposalInfo = obj;
+							proposal.hover = hover;
+							getPriorityHelper.adjustCrossReferencePriority(proposal, prefix + (obj as ComplexType).name);
+						}
+						
+						acceptor.accept(proposal);	
+					}
+				}	
+			}
+		}
+	}
+	
+	def String completeTypeConstructor(String base, ComplexType typ) {
+		val namedParamsOpt = ModelUtils.getAccessorParameters(typ);
+		if(namedParamsOpt.present) {
+			val namedParams = namedParamsOpt.get
+			val proposalString = '''«base»«typ.name»(«FOR param : namedParams SEPARATOR(", ")»«param.name» = «getDummyString(param.type)»«ENDFOR»)'''
+			return proposalString;
+		}
+		if(typ instanceof AnonymousProductType) {
+			val proposalString = '''«base»«typ.name»(«FOR conType : typ.accessorsTypes SEPARATOR(", ")»«getDummyString(conType)»«ENDFOR»)'''
+			return proposalString;
+		}
+		if(typ instanceof Singleton) {
+			return '''«base»«typ.name»()'''
+		}
+		return null;
+	}
+	
+	
+	private def getDummyString(Type obj) {
+		if(obj instanceof ComplexType) {
+			if(obj instanceof SumType) {
+				if(obj.alternatives.empty) {
+					return "";
+				}
+				return completeTypeConstructor(obj.name + ".", obj.alternatives.head);
+			}
+			return completeTypeConstructor("", obj);
+		}
+		if(obj instanceof GeneratedType) {
+			if(obj.name == "string") {
+				return '""';
+			}
+			else if(obj.name == "optional") {
+				return 'none()';
+			}
+			else if(obj.name == "array") {
+				return '[]';
+			}
+			return '';
+		}
+		if(obj instanceof PrimitiveType) {
+			if(obj.name == "bool") {
+				return "false";
+			}
+			if(obj.name === null) {
+				return ""
+			}
+			if(obj.name.contains("int")) {
+				return "0";
+			}
+		}
+		return "";
+	}
```

Comment:
Can this logic be also moved to `GeneratorUtils` where we have getDummyExpression()? Then we have the default/dummy string/expression logic on one place. Might also extract all these into a `DefaultValueProvider` with getDefaultString() and getDefaultExpression() dispatch methods for all the different input types.

Reply:
Also, it is safer to use the typesystem instead of name checks like `obj.name.contains("int")`